### PR TITLE
initial commit for Azure Digital Twins

### PR DESCRIPTION
**Pull request by sankumrav**
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [x] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [ ] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [ ] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [ ] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [ ] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

### CODE REVIEW

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json [hunk 1]
```diff
@@ -0,0 +1,1002 @@
...
+    "application/json"
+  ],
+  "produces": [
+    "application/json"
+  ],
+  "paths": {
+    "/digitaltwins/{twinid}": {
+      "get": {
+        "responses": {
+          "default": {
```
Comment (v-EgorKozhadei) [inline certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:23]: @sankumrav using only default responses with defined schema is not allowed. Please refer to [information about responses](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-submission#define-exact-operation-responses) and to [coding standards](https://learn.microsoft.com/en-us/connectors/custom-connectors/coding-standards#responses).

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json [hunk 2]
```diff
@@ -0,0 +1,1002 @@
...
+                  "type": "string",
+                  "description": "Results From the Twin.",
+                  "title": "Result"
+                }
+              }
+            }
+          }
+        },
+        "summary": "Get Twin by Id",
+        "description": "Pass in the twin id to retrieve its details.",
```
Comment (v-EgorKozhadei) [inline certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:38]: @sankumrav please make sure that all titles, summaries, descriptions are grammatically correct and are free of typographical errors. More information can be found in [English strings recommendations](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-submission#create-quality-english-language-strings).

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json [hunk 3]
```diff
@@ -0,0 +1,1002 @@
...
+            "default": "2020-10-31",
+            "x-ms-visibility": "internal"
+          }
+        ]
+      },
+      "delete": {
+        "responses": {
+          "default": {
+            "description": "default",
+            "schema": {}
```
Comment (v-EgorKozhadei) [inline certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:61]: @sankumrav empty response schemas are not allowed and using only default responses is not allowed.

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json [hunk 4]
```diff
@@ -0,0 +1,1002 @@
...
+            "default": "2020-10-31",
+            "x-ms-visibility": "internal"
+          }
+        ]
+      },
+      "put": {
+        "responses": {
+          "default": {
+            "description": "default",
+            "schema": {}
```
Comment (v-EgorKozhadei) [inline certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:88]: @sankumrav same here.

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json [hunk 5]
```diff
@@ -0,0 +1,1002 @@
...
+            }
+          }
+        ],
+        "description": "Creates a new Twin based on the message body."
+      },
+      "patch": {
+        "responses": {
+          "default": {
+            "description": "default",
+            "schema": {}
```
Comment (v-EgorKozhadei) [inline certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:130]: @sankumrav same here. Please look through all responses and define them according to the links in above comment.

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json [hunk 6]
```diff
@@ -0,0 +1,1002 @@
...
+        "items": {
+          "type": "object",
+          "properties": {
+            "op": {
+              "type": "string",
+              "description": "Request Operation."
+            },
+            "path": {
+              "type": "string",
+              "description": "Path that need to be updated."
```
Comment (v-EgorKozhadei) [inline certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:924]: @sankumrav please make sure that descriptions are grammatically correct and are free of typographical errors.

**Review by v-EgorKozhadei — COMMENTED** [review]
Hello @sankumrav,

Please check comments after the review.
Please take a look at Swagger check. It is failing because of some missing properties and requirements. You can see the details below message section.

Let us know if you have any questions.

File: certified-connectors/AzureDigitalTwins/apiProperties.json [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+          "x-ms-apimTemplateParameter.urlTemplate": "https://@connectionParameters('adtInstance')"
+        }
+      },
+      {
+        "parameters": {
+          "x-ms-apimTemplate-operationName": [
+            "UpdateTwinStringProp",
+            "UpdateTwinBooleanProp"
+          ],
+		      "x-ms-apimTemplateParameter.httpMethod": "@Request.OriginalHTTPMethod",
```
[the remark this answers is not in the document]
Comment (sankumrav) [reply certified-connectors/AzureDigitalTwins/apiProperties.json]: Fixed

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json (hunk 3 above)
Comment (sankumrav) [reply certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:61]: Azure digital twins have dynamic schemas and the response depends on the customer's model. It's not possible to define a schema upfront.

Comment (sankumrav) [reply certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:61]: https://learn.microsoft.com/en-us/connectors/custom-connectors/certification-submission#step-2-meet-submission-requirements 

There is a point where it allows exceptions: "Empty response schemas aren't allowed except in special cases where the response schema is dynamic. This means no dynamic content will be shown in the output and makers must use JSON to parse the response."

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json (hunk 4 above)
Comment (sankumrav) [reply certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json:88]: same as above.

File: certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json (hunk 3 above)
Comment (sankumrav) [reply certified-connectors/AzureDigitalTwins/apiDefinition.swagger.json]: Nevermind. Fixed it.

**Review by v-EgorKozhadei — APPROVED** [review]

**Comment by v-EgorKozhadei** [comment]
Hello @sankumrav, 

I hope you are doing well. 

Congratulations, your pull request is approved and merged. Please [submit the files to ISV Studio](https://docs.microsoft.com/en-us/connectors/custom-connectors/submit-for-certification#submit-to-isv-studio) or [update your submission](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-updates) using your latest commit ID, please allow us up to 1-2 weeks to review your submission, and our Engineers will notify you in the “Activity Control” area in ISV Studio. Please [make sure that the icon meets all the requirements](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-submission#design-an-icon-for-your-connector), that it’s size is 230x230, icon background is not white or transparent and it matches the iconBrandColor property in apiProperties.json file. 

After making the submission in ISV Studio, please attach an intro.md file in the activity control section. This file is required and it's different from readme.md. You can [find an example of intro.md file on this page](https://docs.microsoft.com/en-us/connectors/custom-connectors/submit-for-certification#submit-to-isv-studio) at step 6. 

Please [Create an environment in the Preview region](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-testing#create-an-environment-in-the-preview-region). You will use this environment later to [test your connector](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-testing#test-your-connector) after Microsoft is done with the functional verification for your connector.  

We expect all tests to be completed within 1-2 weeks. 

If your connector passes all tests, the deployment process begins, and it typically takes up to 3 to 4 weeks to deploy your connector to all of our regions. 

Please let me know if you have any questions or run into any problems while creating or updating your submission. 

Thank you very much for working with us.

**Pull request merged**